lib/AST/StructGenerator.cpp

```diff
@@ -452,33 +452,56 @@ clang::RecordDecl* StructGenerator::GetRecordDecl(llvm::DICompositeType* t) {
   return decl;
 }
 
-clang::EnumDecl* StructGenerator::GetEnumDecl(llvm::DICompositeType* t) {
-  auto& decl{enum_decls[t]};
-  if (decl) {
-    return decl;
+clang::QualType StructGenerator::GetEnumDecl(llvm::DICompositeType* t) {
+  auto& type{enum_types[t]};
+  if (!type.isNull()) {
+    return type;
   }
 
   auto tudecl{ast_ctx.getTranslationUnitDecl()};
-  decl = ast.CreateEnumDecl(tudecl, GetUniqueName(t));
   auto base{BuildType(t->getBaseType())};
-  auto i{0U};
-  for (auto elem : t->getElements()) {
-    if (auto enumerator = llvm::dyn_cast<llvm::DIEnumerator>(elem)) {
-      auto elem_name{enumerator->getName().str()};
-      if (elem_name == "") {
-        elem_name = "anon";
+  auto name{GetUniqueName(t)};
+  if (base == ast_ctx.IntTy || base == ast_ctx.UnsignedIntTy) {
+    auto decl{ast.CreateEnumDecl(tudecl, name)};
+    auto i{0U};
+    for (auto elem : t->getElements()) {
+      if (auto enumerator = llvm::dyn_cast<llvm::DIEnumerator>(elem)) {
+        auto elem_name{enumerator->getName().str()};
+        if (elem_name == "") {
+          elem_name = "anon";
+        }
+        elem_name += "_" + std::to_string(i++);
+
+        auto cdecl{ast.CreateEnumConstantDecl(
+            decl, elem_name, ast.CreateIntLit(enumerator->getValue()))};
+        decl->addDecl(cdecl);
       }
-      elem_name += "_" + std::to_string(i++);
+    }
+    decl->completeDefinition(base, base, 0, 0);
 
-      auto cdecl{ast.CreateEnumConstantDecl(
-          decl, elem_name, ast.CreateIntLit(enumerator->getValue()))};
-      decl->addDecl(cdecl);
+    tudecl->addDecl(decl);
+    type = ast_ctx.getEnumType(decl);
+  } else {
+    auto tdef{ast.CreateTypedefDecl(tudecl, name, base)};
+    tudecl->addDecl(tdef);
+    type = ast_ctx.getTypedefType(tdef);
+
+    auto i{0U};
+    for (auto elem : t->getElements()) {
+      if (auto enumerator = llvm::dyn_cast<llvm::DIEnumerator>(elem)) {
+        auto elem_name{enumerator->getName().str()};
+        if (elem_name == "") {
+          elem_name = "anon";
+        }
+        elem_name += "_" + std::to_string(i++);
```

Comment:
I think it would be nice in the future to have a `NamingPolicy` class of some kind that would be responsible to naming things.